### ISSUE TITLE
[react-navigation] unable to use typing for navigation.state.params

### ISSUE THREAD

**Issue by Cherrionella** [flow-typed/flow-typed#2959]
In `react-navigation_v2.x.x` flow types we have type
```
type NavigationParams = {
    [key: string]: mixed,
}
```
So if i define my navigation prop as
`navigation: NavigationScreenProp<NavigationStateRoute>` it assumes that params are of mixed type, but i need exact, like:
```
type Props = {
    navigation: NavigationScreenProp<NavigationStateRoute<{
        foo: number,
        bar: string
    }>>`
}
```

Also it would be nice, if `navigation.setParams()` will have $Shape definition so it can autocomplete and typecheck passed object

So the main problem is, can we somehow tell flow about navigation state params type?

cc @Ashoat

**Comment by Ashoat** [comment]
It's admittedly non-obvious how to type a custom navigation route. Here's how it's done:

```js
type NavProp = NavigationScreenProp<{|
  ...NavigationLeafRoute,
  params: {|
    color: string,
    threadInfo: ThreadInfo,
    setColor: (color: string) => void,
  |},
|}>;
```

> Also it would be nice, if `navigation.setParams()` will have $Shape definition so it can autocomplete and typecheck passed object

PRs welcome! Make sure to send them to the [react-navigation](https://github.com/react-navigation/react-navigation) repo, though. (Give this [`CONTRIBUTING`](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/react-navigation_v2.x.x/CONTRIBUTING.md) doc a read for context.)

**Comment by Cherrionella** [comment]
Sorry for long answer, but unfortunately this approach doesnt work with `withNavigation` HOC

This code:
```
navigation: NavigationScreenProp<{|
    ...NavigationLeafRoute,
    ...$Exact<NavigationState>,
    params: {|
      categoryId?: number,
      listAll?: boolean,
      disableBreadcrumbs?: boolean,
      slice?: string
    |},
  |}>
```
produces following error: `Cannot call `withNavigation` because number [1] is incompatible with mixed [2] in property `state.params.categoryId`. (References: [1] [2])`

**Comment by Ashoat** [comment]
Try `NavigationStateRoute`, and if that works take a look at its definition to see what you were doing wrong above.

**Comment by Cherrionella** [comment]
Unfortunately that doesn't work, same error as above. Moreover, approach that i tried is just a duplication of `NavigationStateRoute` type with extension of `params` object

**Comment by Ashoat** [comment]
Ah, interesting. Perhaps you are on an older version of Flow - can you try upgrading?

**Comment by Ashoat** [comment]
How are you using React Navigation with Flow 0.81? You’re using React Native right? Are you aware each version of React Native is tied to a specific version of Flow?

**Comment by AndrewSouthpaw** [comment]
What really? So given your RN version you can only use a particular version of Flow?... that hasn't been our experience...

**Comment by Ashoat** [comment]
Yes, that’s correct. If you keep your `.flowconfig` updated using `react-native upgrade` or `react-native-git-upgrade`, there is a declaration at the bottom that locks in a specific Flow version. It may turn out that other versions work, but they’re not supported, and in my experience you usually get errors.

**Comment by AndrewSouthpaw** [comment]
TIL. Thanks!

**Comment by schumannd** [comment]
Maybe I undestand these examples wrong, but they are not working for me.

What I am looking for is a way to define types for `navigation.state.params` that a Component expects.

So that flow will notice and error:
a) if the component A tries to access params, that are not supposed to be there.
b) if another component B navigates to component A, and does not provide the expected params.

A way that only solves case a) is

```
type Props = {
  navigation: { state: {params: {par1: string, par2: number}}},
}

export default class ComponentA extends React.Component<Props> {
  render() {
    console.log(this.props.navigation.state.params.par1) // works
    console.log(this.props.navigation.state.params.par3) // throws flow error
}
```

However, if I tried to navigate to this component from another one like so

```this.props.navigation.navigate('ComponentA', {par1: string});```

It does not error, even though `par2` is missing.

This might not be possible to do with flow, as the connection through `react-navigation` might not be easy to extract.
